File: 프로그래머스/3/43162. 네트워크/네트워크.cpp
#include <string>
#include <vector>

using namespace std;

void dfs(vector<vector<int>>& computers, vector<bool>& visited, int cur) {
    
    visited[cur] = true;
    for(int i = 0 ; i < computers[cur].size() ; i++) {
        if(computers[cur][i] == 1)
            if(!visited[i])
                dfs(computers, visited, i);
    }
}

int solution(int n, vector<vector<int>> computers) {
    
    vector<bool> visited(n, false);
    int cnt = 0;
    for(int i = 0 ; i < n ; i++)
        if(!visited[i]) {
            dfs(computers, visited, i);
            cnt++;
        }

    return cnt;
}

File: 프로그래머스/3/43162. 네트워크/네트워크.java
import java.util.*;
import java.util.stream.*;

class Solution {
    public int solution(int n, int[][] computers) {
        
        // 간선 저장
        Map<Integer, List<Integer>> graph = new HashMap<>();
        for(int i = 0 ; i < computers.length ; i++){
            for(int j = 0 ; j < computers[i].length ; j++){
                // 저장
                if(computers[i][j] == 1){
                    graph.computeIfAbsent(i, k -> new ArrayList<>()).add(j);
                }
            }
        }
        
        // dfs 진행
        int answer = 0;
        boolean[] visited = new boolean[n];
        
        for(int i = 0 ; i < n ; i++){
            if(!visited[i]){
                answer++;
                dfs(graph, visited, i);
            }
        }
        
        return answer;
    }
    
    void dfs(Map<Integer, List<Integer>> graph, boolean[] visited, int cur){
        visited[cur] = true;
        
        for(int next : graph.get(cur)){
            if(!visited[next]){
                dfs(graph, visited, next);
            }
        }
    }
}